### PR TITLE
Improve parser error tolerance around parsing incomplete async members.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Helps address some of the issues we see here: https://github.com/dotnet/roslyn/issues/18621

Specific issue is that if hte parser sees "async SomeType" it conservatively treats this as a field called "SomeType with a return type of 'async'".  This is for back-compat as someone might have written that, and we want to support it, no matter how unlikely that code would be.

To accomplish this, the compiler only parses 'async' out as a modifier if it can prove that a new async member is acutally being written, i.e. only if it sees "async SomeType M".  

However, in cases where the code is entirely in error, but the user hasn't finished writing the member, the tree ends up being quite suboptimal.  For example:

```c#
class C
{
     async Task<IFoo<Bar>>
```

Here, the code is entirely in error, but the existing heuristic feels like there is not enough information to be certain it is a method.  We fall back to parsing a field here, which goes completely off the rails.  

The fix here is to detect cases where the code is definitely in error, and to fall back to the much more likely case that this is actually supposed to be the async modifier.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @gafter @dotnet/roslyn-compiler

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/DeclarationNameCompletionProviderTests_NameDeclarationInfoTests.cs [hunk 1]
```diff
@@ -344,14 +344,13 @@ class C
...
         {
-            // Note that the async is not included in the incomplete member syntax
             var markup = @"
 class C
 {
     async int $$
 }
 ";
-            await VerifySymbolKinds(markup, SymbolKind.Field, SymbolKind.Method, SymbolKind.Property);
+            await VerifySymbolKinds(markup, SymbolKind.Method);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/DeclarationNameCompletionProviderTests_NameDeclarationInfoTests.cs]: This result is now more accurate.  If you write "async int" then you're in progress toward writing a method, not a field or property.

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler @jcouv  Can i get a second pair of eyes?

File: src/Compilers/CSharp/Test/Syntax/Parsing/AsyncParsingTests.cs [hunk 2]
```diff
@@ -1305,5 +1446,257 @@ class C
...
+                    }
+                    N(SyntaxKind.CloseBraceToken);
+                }
+                N(SyntaxKind.EndOfFileToken);
+            }
+            EOF();
+        }
     }
-}
+}
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Syntax/Parsing/AsyncParsingTests.cs]: Do you want to also add a unittest that is closer to the end-to-end described in the bug (looking up a symbol in incomplete syntax)?

Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/Parsing/AsyncParsingTests.cs]: Sounds like a good idea :)

**Comment by CyrusNajmabadi** [comment]
@jcouv Have added specific scenario test here.  Are you ok with this change now?

Thanks1

**Comment by CyrusNajmabadi** [comment]
Tagging @ManishJayaswal 

**Customer scenario**

User starts typing an async member that returns a generic task.  That generic task references a type not in scope.  User expects to be able to use "add using" to bring that that type into scope.

**Bugs this fixes:**

https://github.com/dotnet/roslyn/issues/18621

**Workarounds, if any**

None

**Risk**

Low.

**Performance impact**

Low.

**Is this a regression from a previous update?**

No.

**Root cause analysis:**

This is a case of making features work well with incomplete, ambiguous-to-parse, code.  It was just one of those cases where the compiler didn't produce a great tree, thus making down-the-line features do more poorly.

**How was the bug found?**

Customer reported.

**Comment by ManishJayaswal** [comment]
approved pending successful test run.

**Pull request merged**